### ISSUE TITLE
multiple markers on Rails platform

### ISSUE THREAD

**Issue by andrewsunglaekim** [mapbox/mapbox-gl-js#1028]
I'm trying to incorporate markers on a map using Mapbox GL on my rails app.

In my controller file i have:
```
@boatramps = Boatramp.all.limit(6)
@geojson = {"type" => "FeatureCollection", "features" => []}
@boatramps.each do |boatramp|
  @geojson["features"] << {
    geometry: {
    type: 'Point',
    coordinates: [boatramp.long, boatramp.lat]
    },
    properties: {
    title: boatramp.id,
    :'marker-symbol' => 'monument'
    }
  }
end
respond_to do |format|
  format.html
  format.json{render json: @geojson}
end
```
for my ajax call I have
```
style.layers.push({
    "id": "markers",
    "type": "symbol",
    "source": "markers",
    "layout": {
      "icon-image": "{marker-symbol}-12",
      "text-field": "{title}",
      "text-font": "Open Sans Semibold, Arial Unicode MS Bold",
      "text-offset": [0, 0.6],
      "text-anchor": "top"
    },
    "paint": {
      "text-size": 12
    }
  });
$.ajax({
    type: 'GET',
    url: '/virginia',
    dataType: 'json',
    success: function(response){
        console.log(response)
        var markers = new mapboxgl.GeoJSONSource({ data: response});

        map.addSource('markers', markers);
    },
    error: function(){

    }
})
```
When I have 5 markers the call works fine and i get the boatramp id to show.(Also having issues with the "icon image" to show) As soon as I increase the limit to 6 or more. I get an error:

Uncaught TypeError: Cannot read property 'leaf' of undefined

As I zoom in and out on the map it executes the same error. Additionally, once I Have a considerable amount of objects 40 +, some of them don't show. Any ideas?

**Comment by jfirebaugh** [comment]
Can you provide a link to live code that demonstrates the issue?

**Comment by andrewsunglaekim** [comment]
yeah, i’ll try and get a code pen on it, the problem is I’m going to have to hard code the ajax call, but i’ll try and get something to link to

**Comment by andrewsunglaekim** [comment]
This is the codepen i used to emulate it

http://codepen.io/andrewsunglaekim/pen/bNMxOV

Unfortunately, the codepen runs fine. The Ajax call i'm using formats the response object into the FeaturesCollection and runs fine for 5 markers. As soon as I get past 5 it busts.

**Comment by andrewsunglaekim** [comment]
additionally, the markers don't show up, the title does, but not the marker

**Comment by jfirebaugh** [comment]
So this might be an issue in the ajax code rather than mapbox-gl-js? Or can you rule that out?

**Comment by andrewsunglaekim** [comment]
The Ajax code works fine with 5 objects in the features array. As soon as I make it 6 the ajax call still works, I get the objects, but then i get the error

`Uncaught TypeError: Cannot read property 'leaf' of undefined`

And when i zoom in and out it calls the same error multiple times

**Comment by andrewsunglaekim** [comment]
@jfirebaugh I'm getting the object markers on the map, but they only render at certain zoom levels, is there a way to pass in a parameter to specify the min/max zoom levels for each point?

**Comment by jfirebaugh** [comment]
> is there a way to pass in a parameter to specify the min/max zoom levels for each point?

Not on a point-by-point basis, only for the layer as a whole: `minzoom` and `maxzoom`.

"Cannot read property 'leaf' of undefined" sounds like #933, but I think sprite atlas overflow is unlikely to be the cause given that you're using a default style and sprites. Ultimately we're going to need a link to live code that demonstrates the issue to be able to debug this.

**Comment by mourner** [comment]
@jfirebaugh generally it looks like `rbush` is giving this error when you insert invalid bboxes into the tree (e.g. with `NaN` coords).

**Comment by andrewsunglaekim** [comment]
i'm writing this app for a final project, when i take it to "production" ill post a link here ~thursday

**Comment by andrewsunglaekim** [comment]
@jfirebaugh & @mourner heres a link to the live deployed example @[https://stark-mesa-9660.herokuapp.com/virginia](https://stark-mesa-9660.herokuapp.com/virginia)

the main part of the code can be found [here](https://github.com/andrewsunglaekim/paddler_plan/blob/master/app/assets/javascripts/backbone/views/boatramps.js)

i put it together kind of quick to try and get some info to you guys, so bear with the code.

**Comment by andrewsunglaekim** [comment]
if you zoom in on that link, you can start to see the markers in virginia

**Comment by mourner** [comment]
@andreasviglakis is there any way you could reduce the test case to the absolute minimum, without any ajax calls etc., and put it on JSFiddle?

**Comment by andrewsunglaekim** [comment]
heres the thing, on code pen when I try to emulate w/o ajax call its puts 6 markers on fine. The crazy thing is that the code works fine for 5 instances or less, but then runs into zooming issues with more than 6

**Comment by andrewsunglaekim** [comment]
actually the codepen stops working too.. it just takes more instances. So if you look at this one [here](http://codepen.io/andrewsunglaekim/pen/bNMxOV) you can't see the markers near dc until zooming in. This is basically the same issue I'm having. And if you open up the console you can see the 'leaf' error. You can also see that the number of error goes up as you zoom in and out.

**Comment by jfirebaugh** [comment]
I should have remembered: #987. This was fixed in a33d27ec47acbcb62bd35db82f83b49d399f22c7 and will be in the next release.

**Comment by andrewsunglaekim** [comment]
is there any way I can modify my code in the application to change the tileSize locally? Because I link to the mapbox gl in my html.
